pkg/spacebinding/spacebindinglister_test.go

```diff
@@ -0,0 +1,104 @@
+package spacebinding_test
+
+import (
+	"fmt"
+	"testing"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	"github.com/codeready-toolchain/toolchain-common/pkg/spacebinding"
+	"github.com/codeready-toolchain/toolchain-common/pkg/test"
+	. "github.com/codeready-toolchain/toolchain-common/pkg/test/masteruserrecord"
+	spacetest "github.com/codeready-toolchain/toolchain-common/pkg/test/space"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewSpaceBindingLister(t *testing.T) {
+
+	t.Run("recursive list for space", func(t *testing.T) {
+		t.Run("success", func(t *testing.T) {
+			// given
+			space := spacetest.NewSpace(test.HostOperatorNs, "smith",
+				spacetest.WithTierName("advanced"),
+				spacetest.WithSpecTargetCluster(test.MemberClusterName),
+				spacetest.WithLabel(toolchainv1alpha1.SpaceCreatorLabelKey, "johnny"),
+			)
+			subSpace := spacetest.NewSpace(test.HostOperatorNs, "smith-sub",
+				spacetest.WithTierName("advanced"),
+				spacetest.WithSpecTargetCluster(test.MemberClusterName),
+				spacetest.WithLabel(toolchainv1alpha1.SpaceCreatorLabelKey, "johnny"),
+				spacetest.WithSpecParentSpace(space.GetName()),
+			)
+			mur := NewMasterUserRecord(t, "johnny", TargetCluster(test.MemberClusterName), TierName("deactivate90"))
+			listSpaceBindingsFunc := func(spaceName string) ([]toolchainv1alpha1.SpaceBinding, error) {
+				return []toolchainv1alpha1.SpaceBinding{
+					*spacebinding.NewSpaceBinding(mur, space, "johnny", spacebinding.WithRole("admin")),
+				}, nil
+			}
+			getSpaceFunc := func(spaceName string) (*toolchainv1alpha1.Space, error) {
```

Comment:
I would make this test a bit smarter to be closer to the real world and cover more usecases.

First at all I would create at least three layers (so you can fully test the recursion): Parent, two child, and maybe one grandchild.

Then create two smart mock function:

```
getSpaceFunc := func(spaceName string) (*toolchainv1alpha1.Space, error) {
        for _, space := range spaces {
                if space.Name == spaceName {
                        return space, mil
                }
        }
        return nil, nil
}
```
```
listSpaceBindingsFunc := func(spaceName string) ([]toolchainv1alpha1.SpaceBinding, error) {
        space, _ := getSpaceFunc(spaceName)
	return []toolchainv1alpha1.SpaceBinding{
		*spacebinding.NewSpaceBinding(mur, space, spaceName + "-binding", spacebinding.WithRole("admin")),
	}, nil
}
```
Then I would test all three cases when you list SBs for all three layers. You could create a matrix of the given data and given space you want to list SB for and the expected spacebings to avoid code duplication.

Reply:
good point! Plz take a look at https://github.com/codeready-toolchain/toolchain-common/pull/329/commits/ef238f6853844b96929ed97b19bfa678b89a9b54

I couldn't implement `listSpaceBindingsFunc` exactly how you suggested it, since I needed more control over the spacebindigns that are returned. For example I don't always want to return a spacebinding object for every space in the tree, but it was good source of inspiration.

Please let me know if some use-case is missing.